File: src/main/java/algo/palindrome/PalindromicSubstrings_647.java
package algo.palindrome;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

public class PalindromicSubstrings_647 {
  public static int countSubstrings(String s) {
    var dp = new HashMap<Integer, Set<Integer>>();  // start indices of palindrome ends at i
    dp.put(0, new HashSet<>());
    dp.get(0).add(0);
    var count = 1;
    for (var i = 1; i < s.length(); i++) {
      dp.put(i, new HashSet<>());
      dp.get(i).add(i);
      // check if we can make palindrome
      for (var start : dp.get(i - 1)) {
        if (isPalindrome(s, start, i)) {
          dp.get(i).add(start);
        }
        if (start > 0 && s.charAt(start - 1) == s.charAt(i)) {
          dp.get(i).add(start - 1);
        }
      }
      count += dp.get(i).size();
    }
    return count;
  }

  private static boolean isPalindrome(String s, int start, int end) {
    for (int i = start, j = end; i < j; i++, j--) {
      if (s.charAt(i) != s.charAt(j)) {
        return false;
      }
    }
    return true;
  }

  public static void main(String[] args) {
    System.out.println(countSubstrings("abc"));
    System.out.println(countSubstrings("aaa"));
  }
}

File: src/main/kotlin/algo/palindrome/PalindromicSubstrings_647.kt
package algo.palindrome

class KtPalindromicSubstrings_647 {
    fun countSubstrings(s: String): Int {
        // end index of palindrome ends at i -> start indices of all palindromes
        val palindrome: MutableMap<Int, MutableList<Int>> = mutableMapOf()
        val dp = IntArray(s.length)
        s.indices.forEach { i ->
            palindrome.getOrPut(i) { mutableListOf() }.add(i)
            dp[i]++

            // check if we can make palindrome by adding chars around idx to i - 1
            if (i > 0) {
                palindrome[i - 1]?.filter { it > 0 }?.forEach { idx ->
                    if (s[i] == s[idx - 1]) {
                        palindrome[i]?.add(idx - 1)
                        dp[i]++
                    }
                }
            }

            // check if we can make palindrome by adding s[i] to prev palindrome
            palindrome[i - 1]?.forEach check@{ idx ->
                if (palindrome[i]!!.contains(idx)) {
                    return@check
                }
                var isPalindrome = true
                for (k in idx..(idx + (i - idx) / 2)) {
                    if (s[k] != s[i + idx - k]) {
                        isPalindrome = false
                        break
                    }
                }
                if (isPalindrome) {
                    palindrome[i]?.add(idx)
                    dp[i]++
                }
            }
        }
        return dp.sum()
    }
}

fun main() {
    val palindrome = KtPalindromicSubstrings_647()
    println(palindrome.countSubstrings("abc"))
    println(palindrome.countSubstrings("aaa"))
}